<a href="https://colab.research.google.com/github/Alekhyo/Data-Wrangling-EDA/blob/master/EDA1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import pandas as pd
import numpy as np
import math

In [0]:
patients=pd.read_csv('patients.csv')
treatment=pd.read_csv('treatments.csv')
adverse=pd.read_csv('adverse_reactions.csv')

In [0]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

In [0]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [0]:
treatment

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [0]:
adverse

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

Treatment: missing in hba1c change (Completeness)---DONE,
novodra-aurolin is string containing unit (validity)---DONE
data missing(only 280 rows instead of 350) (completeness)---DONE
name and surname in lowercase, (consistency)---DONE
incorrect data type for auralin and novadra, (validity)---DONE
hba1c change 9 instead of 4, (accuracy)---DONE
put nan values in place of hyphen (consistency)---DONE


Patients: Zip code in decimal, (Validity)---DONE
zip code of 4 digits, (Validity)---DONE
shortform of states,(consistency)---DONE
missing contact details, (completeness)----DONE
incorrect data type for assigned_sex-birthdate columns, (?)---DONE
erronous data for wt-height, (Accuracy)---DONE
inconsistent way of storing phone numbers in contact,(consistency)---DONE
duplicated and not useful data (consistency)---DONE

adverse:  name/surname in lowercase (consistency) ---DONE

Structural Data:

patients: phone and email in same col----DONE

treatments:lower/higher and type of treatment---DONE

adverse: not required. Append in treatment --- DONE

In [0]:
patients_copy=patients.copy()
treatment_copy=treatment.copy()
adverse_copy=adverse.copy()
treatments2=pd.read_csv('treatments_cut.csv')

In [0]:
treatment_copy=pd.concat([treatment_copy,treatments2])

In [0]:
treatment_copy['hba1c_change']=treatment_copy['hba1c_start']-treatment_copy['hba1c_end']

In [0]:
treatment_copy=pd.melt(treatment_copy,id_vars=['given_name','surname','hba1c_start','hba1c_end','hba1c_change'],var_name='treatment',value_name='dose')
treatment_copy=treatment_copy[treatment_copy['dose']!='-']
treatment_copy['start'],treatment_copy['end']=treatment_copy['dose'].str.split('-').str
treatment_copy.drop(columns=['dose'],inplace=True)

C:\Users\ALEKHYO\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  This is separate from the ipykernel package so we can avoid doing imports until


In [0]:
treatment_copy['given_name']=treatment_copy['given_name'].str.capitalize()
treatment_copy['surname']=treatment_copy['surname'].str.capitalize()

In [0]:
import re
patients_copy['phone']=patients_copy['contact'].str.findall("\+?\d?[ -]?\(?\d{3}\)?[ -]?\d{3}[ -]?\d{4}")
patients_copy['email']=patients_copy['contact'].str.findall('[a-zA-Z]+@[a-zA-Z]+\.[a-zA-Z]+')
patients_copy['phone']=patients_copy['phone'].str[0]
patients_copy['email']=patients_copy['email'].str[0]
patients_copy.drop(columns=['contact'],inplace=True)

In [0]:
patients_copy['given_name']=patients_copy['given_name'].str.capitalize()
patients_copy['surname']=patients_copy['surname'].str.capitalize()

In [0]:
adverse_copy['given_name']=adverse_copy['given_name'].str.capitalize()
adverse_copy['surname']=adverse_copy['surname'].str.capitalize()

In [0]:
treatment_copy=treatment_copy.merge(adverse_copy,left_on=['given_name','surname'],right_on=['given_name','surname'],how='outer')

In [0]:
patients_copy['country']=patients_copy['country'].fillna('United States')

In [0]:
patients_copy['zip_code']=patients_copy['zip_code'].fillna(0)
patients_copy['zip_code']=patients_copy['zip_code'].astype('int')
patients_copy['zip_code']=patients_copy['zip_code'].astype('str')

In [0]:
def zip_code(num):
    if len(num)==5:
        return num
    else:
        return '0'+num

In [0]:
patients_copy['zip_code']=patients_copy['zip_code'].apply(lambda code: zip_code(code))

In [0]:
patients_copy.drop_duplicates(subset=['address','email'],keep='first',inplace=True)

In [0]:
patients_copy['birthdate']=patients_copy['birthdate'].apply(pd.to_datetime)

In [0]:
patients_copy['assigned_sex']=patients_copy['assigned_sex'].astype('category')

In [0]:
treatment_copy['start']=treatment_copy['start'].apply(lambda start: int(start.split('u')[0]))

In [0]:
treatment_copy['end']=treatment_copy['end'].apply(lambda end: int(end.split('u')[0]))

In [0]:
!pip install uszipcode

In [0]:
from uszipcode import SearchEngine, SimpleZipcode, Zipcode

In [0]:
search=SearchEngine()

In [0]:
def find_state(zip_code):
    zip_code=int(zip_code)
    zipcode = search.by_zipcode(zip_code)
    return zipcode.state

In [0]:
patients_copy['States']=patients_copy['zip_code'].apply(lambda zip_code:find_state(zip_code))

In [0]:
patients_copy.drop(columns='state',inplace=True)

In [0]:
patients_copy=patients_copy[['patient_id', 'assigned_sex', 'given_name', 'surname', 'address',
       'city', 'zip_code', 'States','country', 'birthdate', 'weight', 'height', 'bmi',
       'phone', 'email']]

In [0]:
np.unique(patients_copy['height'].values)

array([27, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74,
       75, 76, 77, 78, 79], dtype=int64)

In [0]:
wt=patients_copy.loc[patients_copy['height']==27,'weight'].values[0]
bmi=patients_copy.loc[patients_copy['height']==27,'bmi'].values[0]
patients_copy.loc[patients_copy['height']==27,'height']=round(math.sqrt((703*wt)/bmi),2)

In [0]:
ht=patients_copy.loc[patients_copy['weight']==48.8,'height'].values[0]
bmi=patients_copy.loc[patients_copy['weight']==48.8,'bmi'].values[0]
patients_copy.loc[patients_copy['weight']==48.8,'weight']=round((bmi*(ht**2)/703),2)

In [0]:
patients_copy.describe()

patient_id      weight      height         bmi
count  484.000000  484.000000  484.000000  484.000000
mean   251.780992  173.379194   66.667293   27.496074
std    147.331198   33.756135    4.045068    5.279599
min      1.000000  102.100000   59.000000   17.100000
25%    122.750000  148.600000   63.000000   23.200000
50%    252.000000  174.450000   67.000000   27.200000
75%    381.250000  199.575000   70.000000   31.700000
max    502.000000  255.900000   79.000000   37.700000

In [0]:
treatment_copy.sample(5)

given_name    surname  hba1c_start  hba1c_end  hba1c_change treatment  \
165      Genet   Fesahaye         7.73       7.36          0.37   auralin   
271      Valur  Bjarkason         9.71       9.41          0.30   novodra   
18         Chỉ        Lâm         7.68       7.24          0.44   auralin   
98       Peter   Pospíšil         7.76       7.34          0.42   auralin   
76       Helen      Luwam         7.66       7.19          0.47   auralin   

     start  end adverse_reaction  
165     25   35              NaN  
271     31   36              NaN  
18      45   48              NaN  
98      51   60              NaN  
76      32   42              NaN

In [0]:
patients_copy.sample(5)

patient_id assigned_sex given_name      surname                  address  \
464         465       female      Meaza       Brhane         142 Broad Street   
399         400         male       Eric           Ek        4160 Pratt Avenue   
153         154         male       John     Carreiro  1463 Martha Ellen Drive   
54           55       female     Louise      Johnson     4984 Hampton Meadows   
371         372       female       Asya  Musliyevich  1763 Williams Mine Road   

            city zip_code States        country  birthdate  weight  height  \
464  Springville    35146     AL  United States 1991-06-22   161.3    69.0   
399      Seattle    98109     WA  United States 1943-04-07   212.1    71.0   
153         Reno    89509     NV  United States 1976-04-08   177.8    68.0   
54    Burlington    01803     MA  United States 1931-03-01   141.0    62.0   
371    Weehawken    07087     NJ  United States 1995-07-15   156.0    63.0   

      bmi              phone                        email  
464  23.8       205-467-9279    MeazaBrhane@superrito.com  
399  29.6  +1 (360) 965-1558        EricEk@jourrapide.com  
153  27.0     1 775 770 7827  JohnACarreiro@superrito.com  
54   25.8       978-407-1874      LouiseJohnson@rhyta.com  
371  27.6  +1 (908) 287-7099  AsyaMusliyevich@armyspy.com